app/Actions/FindPetsForPreferenceAction.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Actions;
+
+use App\Models\Pet;
+use App\Models\Preference;
+use App\Services\PetMatcher;
+use App\Services\ProximityService;
+use Illuminate\Support\Collection;
+
+class FindPetsForPreferenceAction
+{
+    public function __construct(
+        protected ProximityService $proximity,
+        protected PetMatcher $matcher,
+    ) {}
+
+    public function execute(?Preference $preference): Collection
+    {
+        $pets = Pet::with("tags", "shelter.address");
+
+        if ($preference && $preference->latitude && $preference->longitude) {
+            $pets = $this->proximity->withinRadius(
+                $preference->latitude,
+                $preference->longitude,
+                $preference->radius_in_km,
+            );
+        }
+
+        return $pets->get()
+            ->map(function (Pet $pet) use ($preference): array {
+                $petData = $this->normalizePet($pet);
+
+                $matchPercentage = $preference
+                    ? $this->matcher->match($petData, $preference->preferences)
+                    : 0;
+
+                return [
+                    "pet" => $pet,
+                    "match" => $matchPercentage,
+                ];
+            })
+            ->sortByDesc("match")
+            ->values();
+    }
+
+    private function normalizePet(Pet $pet): array
+    {
+        return [
+            "id" => $pet->id,
+            "species" => $pet->species,
+            "tags" => $pet->tags->map(fn($t) => ["id" => $t->id, "name" => $t->name])->all(),
```

Comment:
Please change `$t` to `$tag`, provide type of variable and expected function's return type